src/backends/tensorrt/plan_backend.cc

```diff
@@ -197,6 +193,64 @@ PlanBackend::CreateExecutionContexts(
                                           std::to_string(c) + "_gpu" +
                                           std::to_string(gpu_device);
 
+        // [TODO] abstract the following as a function
+        // Determine the model file to use for device compute capability
+        auto eit = device_engine_.find(gpu_device);
+        if (eit == device_engine_.end()) {
+          eit = device_engine_
+                    .emplace(gpu_device, std::make_pair(nullptr, nullptr))
+                    .first;
+
+          cudaDeviceProp cuprops;
+          auto cuerr = cudaGetDeviceProperties(&cuprops, gpu_device);
+          if (cuerr != cudaSuccess) {
+            return Status(
+                RequestStatusCode::INTERNAL,
+                "unable to get CUDA device properties for " + Name() + ": " +
+                    cudaGetErrorString(cuerr));
+          }
+
+          const std::string cc = std::to_string(cuprops.major) + "." +
+                                 std::to_string(cuprops.minor);
+          const auto& cc_itr = Config().cc_model_filenames().find(cc);
+          const std::string& cc_model_filename =
+              (cc_itr == Config().cc_model_filenames().end())
+                  ? Config().default_model_filename()
+                  : cc_itr->second;
+
+          const auto& mn_itr = models.find(cc_model_filename);
+          if (mn_itr == models.end()) {
+            return Status(
+                RequestStatusCode::INTERNAL, "unable to find PLAN model '" +
+                                                 cc_model_filename + "' for " +
+                                                 Name());
+          }
+
+
+          // Create a CUDA engine shared by all contexts
+          RETURN_IF_ERROR(LoadPlan(
+              mn_itr->second, &eit->second.first, &eit->second.second));
+          // Validate whether the engine can be shared
+          bool is_dynamic = false;
+          auto binding_count = eit->second.second->getNbBindings();
+          for (int idx = 0; idx < binding_count; idx++) {
+            auto dims = eit->second.second->getBindingDimensions(idx);
+            // Detect whether dynamic or not
+            if (ContainsWildcard(dims)) {
+              is_dynamic = true;
+              break;
+            }
+          }
+          // Model with dynamic shapes can't share engine, set to engine to
```

Comment:
Are you sure dynamic shapes can't share engine? As per what I know, the engine can still be shared but multiple contexts can not use a same optimization profile simultaneously.

Reply:
"multiple contexts can not use a same optimization profile simultaneously" is the issue, if we share engine in dynamic shape case, then we need to make sure the instances are aware of which profiles are in used.